Add map region mask

registerrier opened the pull request:
This PR adds a `make_region_mask`method to `HpxNDMap` and `WcsNDMap`.

cdeil commented on gammapy/maps/hpx.py:
```diff
@@ -1731,6 +1732,23 @@ def solid_angle(self):
         return Quantity(hp.nside2pixarea(self.nside), 'sr')
 
+    def get_region_idx(self, region):
+        """Return idx of pixels inside region
```
I don't think this description is correct. It returns a boolean mask, not a pixel idx array, no?

cdeil commented on gammapy/maps/hpxnd.py:
```diff
@@ -648,3 +649,28 @@ def get_angle(x, t):
 
         return fig, ax, p
+
+    def make_region_mask(self, region, inside=True):
```
I think this code is 100% identical in WCS and HPX case.
Suggest to avoid the code duplication and move it to the base class.

registerrier replied:
A priori no because you have no constructor on the base class.
I suppose we can do it using `Map.from_geom()`and then assign `Map.data=mask`

cdeil commented on gammapy/maps/hpxnd.py:
```diff
@@ -648,3 +649,28 @@ def get_angle(x, t):
+        if inside is False:
+            np.logical_not(mask,out=mask)
+
+        # TODO : update meta table to include something about the region used for mask creation?
```
Suggest to remove this TODO.

Adding such book-keeping could be done all throughout Gammapy, but would be more than double of actually writing the actual code and needs a lot of thought. Let's write code first, and then think about such provenance later.

registerrier replied:
OK will remove the TODO. 
I put it to remember that at some point we need to do something here with the metadata.

cdeil commented on gammapy/maps/tests/test_wcsnd.py:
```diff
@@ -378,3 +379,15 @@ def test_coadd_unit():
 
+def test_make_region_mask():
+    from regions import CircleSkyRegion
+    geom = WcsGeom.create(npix=(11,11), binsz=2,
```
Suggest to do fewer pixels, so that printouts are nice if an error occurs.
This is what we had before:
https://github.com/gammapy/gammapy/blob/c209481f42edf675dc5d675cb2fd4ffd1ec9873d/gammapy/image/tests/test_core.py#L384

Another key question is if we return `dtype` of `bool` or `int`. Please add an assert to establish that.
I think `bool` makes sense, but then `write` fails because FITS can't handle it? If so, we could consider changing `write` to convert to `unit8` for `bool` maps.

cdeil commented on gammapy/maps/wcs.py:
```diff
@@ -634,6 +635,30 @@ def solid_angle(self):
+        """
+
+        from regions import PixCoord
+        # TODO : if Pixel Compound regions are taken into account, rather convert to PixelRegion
```
Remove this TODO?
I don't understand it, and it doesn't seem to be needed?

registerrier replied:
Compound regions will be needed for exclusion masks, no?

cdeil commented on gammapy/maps/hpxnd.py:
```diff
@@ -648,3 +649,28 @@ def get_angle(x, t):
+    def make_region_mask(self, region, inside=True):
+        """Create a mask of a given region
+
+        TODO: implement list of region for each axis
```
Concerning this TODO:

How should we handle extra axes?

I'm -1 on trying to define ND regions, that's a huge project.

That leaves the option of making a 2D mask, or the same 2D mask duplicated along the axes.

Here we did the replicated along the energy axis:
http://docs.gammapy.org/dev/_modules/gammapy/cube/core.html#SkyCube.region_mask
I think for `solid_angle` that's also what you did @registerrier ?
I would be in favour of returning 2D by default, but adding a bool option which replicates and makes an ND mask.

And then I'd also suggest to add an example how to do something more fancy, e.g. if there is an energy-dependent FOV one can make an empty mask, then call `iter_by_image` and fill as you like:
http://docs.gammapy.org/dev/maps/index.html#iterating-on-a-map
Inventing an API for that is difficult, and things can be added later, so for now I'd suggest to solve this via docs.

registerrier replied:
Well that was the idea basically. 
You can provide a list/tuple of SkyRegions and then apply iteration.

joleroi commented:
> I don't think at the moment we should support it here, which makes we don't support region filtering for HPX at all.

Agreed, let's leave this to astropy.regions

joleroi commented on gammapy/maps/hpx.py:
```diff
@@ -1731,6 +1732,23 @@ def solid_angle(self):
+            raise NotImplementedError
+
+        coords=self.get_coord()
+        return region.contains(coords.skycoord)
```
Note that in future ``regions`` releases ``contains`` will require a ``wcs`` object (see [latest docs](http://astropy-regions.readthedocs.io/en/latest/api/regions.SkyRegion.html?#regions.SkyRegion.contains)). This is due to the fact, that ``SkyRegions`` are basically ``PixelRegions`` with the coordinates defined on the sky (as opposed to real sky regions). 

So this method should work for ``PixelRegions`` - at the moment you raise a ``NotImplementedError`` on that. For convenience you could accept ``SkyRegion`` plus a ``wcs`` object. But like it is now this method will stop working soon.

joleroi commented on gammapy/maps/hpxnd.py:
```diff
@@ -648,3 +649,28 @@ def get_angle(x, t):
+        ----------
+        region :  `~regions.SkyRegion` object
+            A region on the sky could be defined in pixel or sky coordinates.
+            Note that `~regions.PixelRegion` does not work for healpix maps.
```
As mentioned by @cdeil we cannot really support HPX at the moment

registerrier replied:
OK will remove it then.

registerrier commented:
The new commits have:
- removed the region filtering in `map.HpxGeom` and its test function
- changed the name of the method to`WcsGeom.get_region_mask_array`
- include the new calling scheme for `SkyRegion.contains`
- added a `MapCoord.apply_region_mask` and its associated test

Merge?

joleroi commented on gammapy/maps/wcsnd.py:
```diff
@@ -508,3 +508,27 @@ def plot(self, ax=None, idx=None, **kwargs):
+        mask_map : `~gammapy.maps.WcsNDMap`
+            the mask map
+        """
+        mask = self.geom.get_region_mask_array(region)
```
Can't you pass `inside` to `geom.get_region_mask_array` in order to prevent the duplicated check for ``if inside`` in the following lines?

registerrier replied:
OK.

joleroi commented:
> This is a serious issue with regionsno? A SkyRegion should be a region on the sphere no? Otherwise you end up with different selections depending on the WCS object you pass...

Yes, to my understanding this is intended for ``SkyRegion``. It is really a region on some kind of sky projection. The plan is to add ``SphericalSkyRegion`` later, to cover the use case were people want to put a region on an actual sphere. ``SphericalSkyRegion`` will implement circles and polygons. There is some discussion on this in https://github.com/astropy/regions/issues/107

cdeil commented on gammapy/maps/hpx.py:
```diff
@@ -10,6 +10,7 @@
 from astropy.io import fits
 from astropy.coordinates import SkyCoord
 from astropy.units import Quantity
+from regions import SkyRegion
```
Remove unused import

registerrier replied:
Done

cdeil commented on gammapy/maps/hpxnd.py:
```diff
@@ -4,6 +4,7 @@
 import numpy as np
 from astropy.io import fits
 from astropy.units import Quantity
+from regions import SkyRegion
```
Remove unused import

registerrier replied:
Done

cdeil commented on gammapy/maps/tests/test_hpxmap.py:
```diff
@@ -6,6 +6,7 @@
 from astropy.io import fits
 from astropy.coordinates import SkyCoord
 from astropy.units import Quantity
+import astropy.units as u
```
Remove unused import

registerrier replied:
Done

cdeil commented on gammapy/maps/wcs.py:
```diff
@@ -634,6 +635,40 @@ def solid_angle(self):
+    def get_region_mask_array(self, region):
+        """Return mask of pixels inside region in the the form of boolean array
+
+        TODO: implement list of region for each axis
```
I think in my last comment I suggested to remove this TODO (and also in the code below) and to instead show how to do this via a docs example.

registerrier replied:
Well it is not so easy to do actually. I will add an issue about this.

cdeil commented on gammapy/maps/wcs.py:
```diff
@@ -634,6 +635,40 @@ def solid_angle(self):
+        # if isinstance(region, SkyRegion):
+        #    region = region.to_pixel(self.wcs)
+
+        if isinstance(region, SkyRegion):
```
I think this would be simpler and equivalent to how regions works now:
```
if isinstance(region, SkyRegion):
    region = region.to_pixel(self.wcs)

pix = PixCoord(*self.get_idx())
return region.contains(pix)
```

registerrier replied:
Done

cdeil commented on gammapy/maps/tests/test_wcsnd.py:
```diff
@@ -378,3 +379,15 @@ def test_coadd_unit():
+    m = WcsNDMap(geom)
+    region = CircleSkyRegion(SkyCoord(0, 0, unit='deg', frame='galactic'), 1.0*u.deg)
+    maskmap = m.make_region_mask(region)
+    assert np.sum(maskmap.data) == 1
```
Suggest to add an assert on the dtype of data. It's bool, no?

registerrier replied:
Done

registerrier commented:
OK I have made the modifications. 
There are some difficulties which I had not foreseen with multi resolution maps. I have added an Exception if one try to use one. This will need to be sorted out later.
I have also removed for the moment the MapCoord region filtering. This might be better handled with a MapMask for instance.
